canonical-web-and-design/maas-ui
# feat(ui): add take action menu to machine details

## Done

- Moved machine list action forms into `app/machines/components`
- Refactored machine list action forms so that they could also be used in machine details.
- Made sure that each form loads the necessary data on load, and updated `ActionForm` to show a spinner accordingly.
- Added `useMachineActionForm` hook which reduces a lot of the commonalities between the machine list/machine details forms.
- Standardised machine actions so that they all accept machine system IDs.
- Updated `machineSelectors.fetchFailedScriptResults` to accept a list of machine IDs instead of only using the selected IDs in state
- Fixed a bug where the "Power on" action would not close the form on completion.

## QA

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Go to the machine details page of a machine and refresh the page so state has not been populated from the machine list
- Check that each of the forms works as expected:
- [x] Commission
- [x] Acquire
- [x] Abort
- [x] Power on
- [x] Power off
- [x] Lock
- [x] Unlock
- [x] Rescue mode
- [x] Exit rescue mode
- [x] Mark broken
- [x] Mark fixed
- [x] Override failed testing
- [x] Set zone
- [x] Set resource pool

## Fixes

Fixes canonical-web-and-design/MAAS-squad#2105
